programs/gummyroll_crud/src/lib.rs

```diff
@@ -4,51 +4,149 @@ use gummyroll::{program::Gummyroll, Node};
+}
+
+#[derive(Accounts)]
+pub struct Add<'info> {
+    #[account(mut)]
+    pub authority: Signer<'info>,
+    #[account(
+        seeds = [
+            b"gummyroll-crud-authority-pda",
+            merkle_roll.key().as_ref(),
+            authority.key().as_ref(),
+        ],
+        bump,
+    )]
+    /// CHECK: This account is neither written to nor read from.
+    pub authority_pda: UncheckedAccount<'info>,
     pub gummyroll_program: Program<'info, Gummyroll>,
+    #[account(mut)]
+    /// CHECK: unsafe
+    pub merkle_roll: UncheckedAccount<'info>,
 }
 
 #[derive(Accounts)]
 pub struct Remove<'info> {
+    #[account(mut)]
+    pub authority: Signer<'info>,
+    #[account(
+        seeds = [
+            b"gummyroll-crud-authority-pda",
+            merkle_roll.key().as_ref(),
+            authority.key().as_ref(),
+        ],
+        bump,
+    )]
+    /// CHECK: This account is neither written to nor read from.
+    pub authority_pda: UncheckedAccount<'info>,
+    pub gummyroll_program: Program<'info, Gummyroll>,
     #[account(mut)]
     /// CHECK: unsafe
     pub merkle_roll: UncheckedAccount<'info>,
-    pub owner: Signer<'info>,
-    pub gummyroll_program: Program<'info, Gummyroll>,
 }
 
 #[derive(Accounts)]
 pub struct Transfer<'info> {
+    /// CHECK: This account is neither written to nor read from.
+    pub authority: UncheckedAccount<'info>,
+    #[account(
+        seeds = [
+            b"gummyroll-crud-authority-pda",
+            merkle_roll.key().as_ref(),
+            authority.key().as_ref(),
+        ],
+        bump,
+    )]
+    /// CHECK: This account is neither written to nor read from.
+    pub authority_pda: UncheckedAccount<'info>,
+    pub gummyroll_program: Program<'info, Gummyroll>,
     #[account(mut)]
     /// CHECK: unsafe
     pub merkle_roll: UncheckedAccount<'info>,
+    #[account(mut)]
     pub owner: Signer<'info>,
-    /// CHECK: This is safe because we don't read from or write to this account.
+    /// CHECK: This account is neither written to nor read from.
     pub new_owner: UncheckedAccount<'info>,
-    pub gummyroll_program: Program<'info, Gummyroll>,
 }
 
 #[program]
 pub mod gummyroll_crud {
 
     use super::*;
 
+    pub fn create_tree(
+        ctx: Context<CreateTree>,
+        max_depth: u32,
+        max_buffer_size: u32,
+    ) -> Result<()> {
+        let gummyroll_program = ctx.accounts.gummyroll_program.to_account_info();
+        let merkle_roll = ctx.accounts.merkle_roll.to_account_info();
+        let authority = ctx.accounts.authority.to_account_info();
+        let authority_pda = ctx.accounts.authority_pda.to_account_info();
+        let authority_pda_bump_seed = &[*ctx.bumps.get("authority_pda").unwrap()];
+        let seeds = &[
+            b"gummyroll-crud-authority-pda",
+            merkle_roll.key.as_ref(),
+            authority.key.as_ref(),
+            authority_pda_bump_seed,
+        ];
+        let authority_pda_signer = &[&seeds[..]];
```

Comment:
I hate all this duplication – here and in the Anchor `seeds` macro – but my Rust is not strong enough to figure out how to DRY this up. Thoughts?

Reply:
Yeah this triggers my mild OCD as well, but I've become a believer in "all code that works is good code." This is what working at a trading firm does to you...